linker/Linker.Steps/RemoveFeaturesStep.cs

```diff
@@ -17,79 +18,103 @@ protected override void ProcessAssembly (AssemblyDefinition assembly)
 		{
-			RemoveCustomAttributes (type);
+			RemoveCom (type);
 
 			foreach (var field in type.Fields)
-				RemoveCustomAttributes (field);
+				RemoveCom (field);
 
 			foreach (var method in type.Methods) {
-				RemoveCustomAttributes (method);
+				RemoveCom (method);
 			}
 
 			foreach (var nested in type.NestedTypes)
 				ProcessType (nested);
 		}
 
-		void RemoveCustomAttributes (ICustomAttributeProvider provider)
+		void RemoveCom (TypeDefinition type)
+		{
+			if (RemoveCustomAttributes (type, IsComAttributeExcluded)) {
+				type.IsImport = false;
+				type.IsWindowsRuntime = false;
+			}
+		}
+
+		void RemoveCom (ICustomAttributeProvider provider)
+		{
+			RemoveCustomAttributes (provider, IsComAttributeExcluded);
+		}
+
+		bool RemoveCustomAttributes (ICustomAttributeProvider provider, Func<CustomAttribute, bool> filter)
 		{
 			if (!provider.HasCustomAttributes)
-				return;
+				return false;
 
-			var attrsToRemove = provider.CustomAttributes.Where (IsCustomAttributeExcluded).ToArray ();
+			var attrsToRemove = provider.CustomAttributes.Where (filter).ToArray ();
 			foreach (var remove in attrsToRemove)
 				provider.CustomAttributes.Remove (remove);
+
+			return attrsToRemove.Length > 0;
 		}
+		
+		bool IsComAttributeExcluded (CustomAttribute attr)
+		{
+			if (FeatureCOM && IsComAttribute (attr))
+				return !IsAttributeTypePreserved (attr);
 
-		bool IsCustomAttributeExcluded (CustomAttribute attr)
+			return false;
+		}
+
+		bool IsAttributeTypePreserved (CustomAttribute attr)
 		{
 			var type = attr.AttributeType;
-
-			switch (type.Name) {
-			default:
+			
+			var definition = type.Resolve ();
+			if (!Annotations.IsPreserved (definition))
 				return false;
 
-			case "ComDefaultInterfaceAttribute":
-			case "ComVisibleAttribute":
-			case "ClassInterfaceAttribute":
-			case "InterfaceTypeAttribute":
-			case "DispIdAttribute":
-			case "TypeLibImportClassAttribute":
-			case "ComRegisterFunctionAttribute":
-			case "ComUnregisterFunctionAttribute":
-			case "ProgIdAttribute":
-			case "ImportedFromTypeLibAttribute":
-			case "IDispatchImplAttribute":
-			case "ComSourceInterfacesAttribute":
-			case "ComConversionLossAttribute":
-			case "TypeLibTypeAttribute":
-			case "TypeLibFuncAttribute":
-			case "TypeLibVarAttribute":
-			case "ComImportAttribute":
-			case "GuidAttribute":
-			case "ComAliasNameAttribute":
-			case "AutomationProxyAttribute":
-			case "PrimaryInteropAssemblyAttribute":
-			case "CoClassAttribute":
-			case "ComEventInterfaceAttribute":
-			case "TypeLibVersionAttribute":
-			case "ComCompatibleVersionAttribute":
-			case "SetWin32ContextInIDispatchAttribute":
-			case "ManagedToNativeComInteropStubAttribute":
-				if (!FeatureCOM || type.Namespace != "System.Runtime.InteropServices")
-					return false;
+			return Annotations.GetPreserve (definition) == TypePreserve.All;
+		}
 
-				break;
-			}
+		public static bool IsComAttribute (CustomAttribute attr)
```

Comment:
Is this really necessary to make it public and have multiple switch statements for each feature in the future?

Reply:
I don't know if we are going to stick with using the `RemoveFeatureStep` or not yet, but in case we didn't, I thought it would be nice to have a helper I could use for checking if a CustomAttribute is COM related.

Not a big deal though.  I can just copy the switch if we don't use `RemoveFeatureStep`